### PR TITLE
feat: transaction access list option

### PR DESCRIPTION
**Pull request by minhd-vu**
## Motivation

- Sometimes when using `cast send` or `cast mktx`, it would be nice to be able to provide an access list

## Solution

- Adds `--access-list` flag for transactions
- `--access-list` provided with no value will use `cast access-list` to populate the access list

```bash
--access-list '[
  {
    "address": "0x528e26b25a34a4a5d0dbda1d57d318153d2ed582",
    "storageKeys": [
      "0xb53127684a568b3173ae13b9f8a6016e243e63b6e8ee1178d6a717850b5d6103",
      "0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc"
    ]
  },
  {
    "address": "0x1f4479ccb333fcba58e68f04e38f3ab070983b31",
    "storageKeys": []
  }
]'
```

### CODE REVIEW

File: crates/cast/bin/tx.rs [hunk 1]
```diff
@@ -172,6 +172,10 @@ where
             if let Some(priority_fee) = tx_opts.priority_gas_price {
                 tx.set_max_priority_fee_per_gas(priority_fee.to());
             }
+
+            if let Some(access_list) = tx_opts.access_list {
```
Comment (klkvr) [inline crates/cast/bin/tx.rs:176]: there is 2930 legacy transaction with access list, so we can set it even if `legacy` is true

File: crates/cli/src/utils/mod.rs [hunk 2]
```diff
@@ -150,6 +152,13 @@ pub fn parse_delay(delay: &str) -> Result<Duration> {
     Ok(delay)
 }
 
+// Parses a `AccessList` from a &str
+pub fn parse_access_list(s: &str) -> Result<AccessList> {
```
Comment (klkvr) [inline crates/cli/src/utils/mod.rs:156]: hmm, I am wondering if we should add some additional formats for better UX here? eg fallback to comma-separated addresses list or something similar

Comment (minhd-vu) [reply crates/cli/src/utils/mod.rs:156]: I like that idea, I'll try my best to implement it.

Comment (klkvr) [reply crates/cli/src/utils/mod.rs:156]: let me know if any questions!

File: crates/cli/src/opts/transaction.rs [hunk 3]
```diff
@@ -62,6 +62,12 @@ pub struct TransactionOpts {
     /// Can be either a hex-encoded signed authorization or an address.
     #[arg(long, conflicts_with_all = &["legacy", "blob"])]
     pub auth: Option<String>,
+
+    /// EIP-2930 access list.
+    ///
+    /// Can be either JSON encoded access list or empty to use the output of `cast access-list`.
```
Comment (mattsse) [inline crates/cli/src/opts/transaction.rs:68]: this probably mention `eth_createAccessList`

File: crates/cli/src/utils/mod.rs [hunk 4]
```diff
@@ -150,6 +152,13 @@ pub fn parse_delay(delay: &str) -> Result<Duration> {
     Ok(delay)
 }
 
+/// Parses an `AccessList` from a &str
+pub fn parse_access_list(s: &str) -> Result<AccessList> {
```
Comment (mattsse) [inline crates/cli/src/utils/mod.rs:156]: if we only support json, then this function is meaningless

File: crates/cast/bin/tx.rs [hunk 5]
```diff
@@ -182,6 +183,20 @@ where
             tx.set_nonce(nonce.to());
         }
 
+        if let Some(access_list) = match tx_opts.access_list {
```
Comment (klkvr) [inline crates/cast/bin/tx.rs:186]: we should do it right before estimating gas in `_build`

we don't yet have calldata here, so calling rpc will result in incomplete data or an error

**Review by klkvr — CHANGES_REQUESTED** [review]

**Review by klkvr — APPROVED** [review]

**Review by mattsse — APPROVED** [review]

**Pull request merged**